python/ray/util/ray_executor.py

```diff
@@ -0,0 +1,237 @@
+import itertools
+import time
+from concurrent.futures import Executor
+from functools import partial
+
+import ray
+from ray.util.annotations import PublicAPI
+
+
+@PublicAPI(stability="alpha")
+class RayExecutor(Executor):
+    """`RayExecutor` is a drop-in replacement for `ProcessPoolExecutor` and
+    `ThreadPoolExecutor` from `concurrent.futures` but distributes and executes
+    the specified tasks over a Ray cluster instead of multiple processes or
+    threads.
+    """
+
+    def __init__(self, max_workers=None, **kwargs):
+        """Initialises a new RayExecutor instance which distributes tasks over
+        a Ray cluster.
+
+        Args:
+            max_workers:
+                The maximum number of Ray workers to distribute the
+                tasks over. This is achieved by creating a fixed pool
+                of Actors (see
+                https://docs.ray.io/en/latest/ray-core/actors/
+                actor-utils.html#actor-pool).
+
+            All additional keyword arguments are passed to ray.init()
+            (see https://docs.ray.io/en/latest/ray-core/package-ref.html#ray-init).
+
+            For example, this will connect to a cluster at the specified address:
+            .. code-block:: python
+
+                RayExecutor(address='192.168.0.123:25001')
+
+            Note: excluding an address will initialise a local Ray cluster.
+        """
+        self._shutdown_lock = False
+        self._futures = []
+        self.__actor_pool = None
+        self.__remote_fn = None
+        self.context = None
+
+        """
+        The following is necessary because `@ray.remote` is only available at runtime.
+        """
+
+        if max_workers is None:
+
+            @ray.remote
+            def remote_fn(fn):
+                return fn()
+
+            self.__remote_fn = remote_fn
+        else:
+            if max_workers < 1:
+                raise ValueError("max_workers must be >= 1")
+
+            @ray.remote
+            class ExecutorActor:
+                def __init__(self):
+                    pass
+
+                def actor_function(self, fn):
+                    return fn()
+
+            actors = [
+                ExecutorActor.options(name=f"actor-{i}").remote()
+                for i in range(max_workers)
+            ]
+            self.__actor_pool = ray.util.ActorPool(actors)
```

Comment:
Is this blocked until all actors are initialized? I think we should either have `ready` API or make sure when the constructor returns all actors are created. Otherwise, it can cause random hanging that's unexpected

Reply:
I have done so (if I understand your suggestion) :)